addons/resource/models/resource.py

```diff
@@ -337,61 +337,100 @@ def _attendance_intervals(self, start_dt, end_dt, resource=None, domain=None, tz
+            for resource in resources_list:
+                if attendance.resource_id.id not in [False, resource.id]:
+                    continue
+                # express all dates and times in specified tz or in the resource's timezone
+                tz = tz if tz else timezone((resource or self).tz)
+                if start_dt in cache_dates[tz]:
+                    start_dt = cache_dates[tz][start_dt]
+                else:
+                    start_dt = start_dt.astimezone(tz)
+                    cache_dates[tz][start_dt] = start_dt
+                if end_dt in cache_dates[tz]:
+                    end_dt = cache_dates[tz][end_dt]
+                else:
+                    end_dt = end_dt.astimezone(tz)
+                    cache_dates[tz][end_dt] = end_dt
+
+                start = start_dt.date()
+                if attendance.date_from:
+                    start = max(start, attendance.date_from)
+                until = end_dt.date()
+                if attendance.date_to:
+                    until = min(until, attendance.date_to)
+                if attendance.week_type:
+                    start_week_type = int(math.floor((start.toordinal()-1)/7) % 2)
+                    if start_week_type != int(attendance.week_type):
+                        # start must be the week of the attendance
+                        # if it's not the case, we must remove one week
+                        start = start + relativedelta(weeks=-1)
+                weekday = int(attendance.dayofweek)
+
+                if self.two_weeks_calendar and attendance.week_type:
+                    days = rrule(WEEKLY, start, interval=2, until=until, byweekday=weekday)
+                else:
+                    days = rrule(DAILY, start, until=until, byweekday=weekday)
+
+                for day in days:
+                    # attendance hours are interpreted in the resource's timezone
+                    if tz not in cache_deltas[day]:
+                        cache_deltas[day][tz] = {}
+
+                    hour_from = attendance.hour_from
+                    if hour_from in cache_deltas[day][tz]:
+                        dt0 = cache_deltas[day][tz][hour_from]
+                    else:
+                        dt0 = tz.localize(combine(day, float_to_time(hour_from)))
+                        cache_deltas[day][tz][hour_from] = dt0
+
+                    hour_to = attendance.hour_to
+                    if hour_to in cache_deltas[day][tz]:
+                        dt1 = cache_deltas[day][tz][hour_to]
+                    else:
+                        dt1 = tz.localize(combine(day, float_to_time(hour_to)))
+                        cache_deltas[day][tz][hour_to] = dt1
+                    result[resource.id].append((max(start_dt, dt0), min(end_dt, dt1), attendance))
+
+        if len(resources) <= 1: # 1 resource or None
+            return Intervals(result[resources.id])
+        return {r.id: Intervals(result[r.id]) for r in resources}
```

Comment:
This is guaranteed future work for the Bugfix team.
Simply call the method with a recordset, you expect a list, and you have an element instead.
Worst API change ever, IMHO.
@nim-odoo

Reply:
I can suggest this API change that would not break everything:
```python
# single resource
result = self._attendance_intervals(..., resource=resource, ...)
assert isinstance(result, Intervals)

# multiple resources: new optional parameter
result = self._attendance_intervals(..., resources=resources, ...)
assert isinstance(result, dict)
```

Reply:
I intend to have 2 methods instead. 1 that always works in batch, and 1 retro-compatible that calls the batched method.

Reply:
> This is guaranteed future work for the Bugfix team.
> Simply call the method with a recordset, you expect a dict, and you have an `Intervals` instead.
> Worst API change ever, IMHO.
> @nim-odoo

Yes, please make sure to have a consistent return type. API changes are tolerated* for private methods, but for sure having multiple return types is a guarantee of having to fix call later on.

* Tolerated in the sense that it's always better to avoid if possible.